casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala

```diff
@@ -49,28 +49,32 @@ object InterpreterUtil {
       _                  <- Span[F].mark("before-compute-parents-post-state")
       preStateHashEither <- computeParentsPostState(parents, dag, runtimeManager).attempt
       _                  <- Log[F].info(s"Computed parents post state for ${PrettyPrinter.buildString(b)}.")
-      invalidBlocksSet   <- dag.invalidBlocks
-      unseenBlocksSet    <- ProtoUtil.unseenBlockHashes(dag, b)
-      seenInvalidBlocksSet = invalidBlocksSet.filterNot(
-        block => unseenBlocksSet.contains(block.blockHash)
-      ) // TODO: Write test in which switching this to .filter makes it fail
-      invalidBlocks = seenInvalidBlocksSet.map(block => (block.blockHash, block.sender)).toMap
-      _             <- Span[F].mark("before-process-pre-state-hash")
-      isGenesis     = b.header.parentsHashList.isEmpty
       result <- preStateHashEither match {
                  case Left(ex) =>
                    BlockStatus.exception(ex).asLeft[Option[StateHash]].pure
                  case Right(computedPreStateHash) =>
                    if (incomingPreStateHash == computedPreStateHash) {
-                     replayBlockDeploys(
-                       runtimeManager,
-                       incomingPreStateHash,
-                       tsHash,
-                       internalDeploys,
-                       BlockData(timestamp, blockNumber),
-                       invalidBlocks,
-                       isGenesis
-                     )
+                     for {
+                       invalidBlocksSet <- dag.invalidBlocks
+                       unseenBlocksSet  <- ProtoUtil.unseenBlockHashes(dag, b)
+                       seenInvalidBlocksSet = invalidBlocksSet.filterNot(
+                         block => unseenBlocksSet.contains(block.blockHash)
+                       ) // TODO: Write test in which switching this to .filter makes it fail
+                       invalidBlocks = seenInvalidBlocksSet
+                         .map(block => (block.blockHash, block.sender))
+                         .toMap
+                       _         <- Span[F].mark("before-process-pre-state-hash")
+                       isGenesis = b.header.parentsHashList.isEmpty
+                       result <- replayBlockDeploys(
+                                  runtimeManager,
+                                  incomingPreStateHash,
+                                  tsHash,
+                                  internalDeploys,
+                                  BlockData(timestamp, blockNumber),
+                                  invalidBlocks,
+                                  isGenesis
+                                )
+                     } yield result
                    } else {
                      //TODO at this point we may just as well terminate the replay, there's no way it will succeed.
```

Comment:
Does this mean that this todo can be removed?

Reply:
I was rather intending to terminate the replay immediately, by raising an error, after figuring out the error handling scheme. I guess @KentShikama should also confirm our understanding. @KentShikama are there any reasons to continue with the replay if the replay of parant states diverged already? (see the TODO in question)

Reply:
I'm confused where this TODO came from. We currently are just returning after logging the warning right?

Reply:
yes. The point of this TODO is: whatever happens after this point is doomed to lead to a faulty replay. In fact, we could proclaim a replay failure right away and slash the block, as clearly the proposing validator confabulates w.r.t. the pre-state hash resulting from merging the parents.

Reply:
That's why I think it'd make sense to raise an error here. The only reasons why I'm not doing that right away are in my first comment.

Reply:
What is happening right now though?

Reply:
After the warning? As mentioned, I think the replay continues but inevitably fails.

Reply:
Yeah. How does the replay continue? If it is, then I agree it shouldn't be.